include/envoy/server/instance.h

```diff
@@ -213,6 +213,11 @@ class Instance {
    * @return the flush interval of stats sinks.
    */
   virtual std::chrono::milliseconds statsFlushInterval() const PURE;
+
+  /**
+   * @return Http::CodeStats the http response-code stats
+   */
+  virtual Http::CodeStats& codeStats() PURE;
```

Comment:
I'm not crazy about this being a server-wide thing and then having to be plumbed through everywhere. What's the reasoning for that? Can it potentially be per-HCM and then just exposed via HTTP filter callbacks? If you want it to be global can we create it in the singleton manager as part of the HCM (like the HTTP date provider)?

Reply:
HCM==HttpConnectionManager I assume?  How many HCMs are there?  One per thread?

The reasoning for having it be global is that it's not going to be free to create or store, once SymbolTables are integrated. But if there were a bounded number created at startup it would be OK. There's no compelling reason there has to be only one, as long as they are not created on demand in the hot-path.

What are the drawbacks having them be plumbed everywhere? In my mind the most painful thing is the number of many edits that required. However, I'd be up for a refactor -- even in advance of this PR. One thought I had was making Server::Configuration::FactoryContext more broadly available, and that would make TimeSource, Http::CodeStats, SymbolTable, and whatever else was needed easier to get.

I hadn't really looked at HCM before but it looks like this is a way to pass around data that's derived from configuration, whereas Server::Configuration::FactoryContext looks more like it's sharing common data structures created by the server at startup, which feels like what this is.

Reply:
> What are the drawbacks having them be plumbed everywhere?

Because IMO it's an abstraction leakage. Envoy is not a dedicated HTTP server (though yes we always have the admin server) so it seems incorrect to create this thing and pass it around everywhere when it only applies to HTTP traffic.

IMO we can create one of these things per HCM (it's not per-thread, it's per listener effectively) or if we want one for all listeners you can create a singleton like here: https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/network/http_connection_manager/config.cc#L66

Reply:
The singleton manager avoids statics, it's an actual object. Take a look at the implementation.

Reply:
I don't think, you're going to be much happier with the change I just made; I was trying to move stuff around for a cleaner dependency graph but I think it was better before I did this.

I just did a commit to see how it looks via the GH interface.  I'll follow up further.

Reply:
Perr DM on slack I am adding a new Http::Context object to hold the httpTracer and Http::CodeStats.

The singleton-manager didn't work out due to same-thread assertions.